File: rl/mcts/mcts_node.py
"""!
@file mcts_node.py
@brief MCTS Node implementation for the Beluga Challenge

This module implements the Monte Carlo Tree Search node structure
for choosing the best parameters for complex actions in the Beluga Challenge.
"""

import math
from rl.env.state import ProblemState  # Assuming State is defined in state.py

# List of complex actions that require parameter selection
complex_actions = ["left_stack_rack", "right_stack_rack", "left_unstack_rack", "right_unstack_rack"]

class MCTSNode:
    """!
    @brief Node class for Monte Carlo Tree Search
    
    Each node represents a state in the search tree and maintains
    information about visits, rewards, and possible actions.
    """
    
    def __init__(self, state, parent=None, action=None, depth=0):
        """!
        @brief Initialize a new MCTS node
        @param state Current problem state
        @param parent Parent node in the search tree
        @param action Action taken to reach this node
        @param depth Depth of this node in the tree
        """
        self.state = state # Current state (later observation)
        self.parent = parent # Parent node
        self.action = action # Action taken to reach this node
        self.depth = depth # Depth in the tree
        self.children = [] # List of child nodes
        self.visits = 0 # Number of visits to this node
        self.total_reward = 0.0 # Total reward accumulated from this node

    def is_root(self):
        """!
        @brief Check if this node is the root of the tree
        @return True if this is the root node, False otherwise
        """
        return self.parent is None

    def is_terminal(self):
        """!
        @brief Check if this node represents a terminal state
        @return True if this is a terminal state, False otherwise
        """
        return self.state.is_terminal()

    def is_fully_expanded(self):
        """!
        @brief Check if all possible actions have been tried from this node
        @return True if fully expanded, False otherwise
        """
        return len(self.get_untried_action()) == 0
        
    def get_untried_action(self):
        """!
        @brief Get list of actions that haven't been tried yet from this node
        @return List of untried (action_name, parameters) tuples
        """
        # If root node has a specific action without parameters
        if self.is_root() and self.action is not None and self.action[1] is None:
            # Only return parameters for this specific action
            action_name = self.action[0]
            all_params = self.state.enumerate_valid_params(action_name)
            tried_params = [child.action[1] for child in self.children]
            # Return only untried parameters
            return [(action_name, param) for param in all_params if param not in tried_params]
        else:
            # Normal behavior for other nodes
            all_possible_actions = self.state.get_possible_actions()
            tried_actions = [child.action for child in self.children]
            return [action for action in all_possible_actions if action not in tried_actions]

    def expand(self, candidate: tuple):
        """!
        @brief Expand the node by adding a new child for the given action
        @param candidate Tuple of (action_name, parameters) to expand
        @return The newly created child node
        """
        new_state = self.state.copy()  # Create copy
        new_state.apply_action(candidate[0], candidate[1])  # Apply action on copy
        child_node = MCTSNode(state=new_state, parent=self, action=candidate, depth=self.depth + 1)
        self.add_child(child_node)
        return child_node

    def add_child(self, child):
        """!
        @brief Add a child node to this node
        @param child Child node to add
        """
        self.children.append(child)

    def best_child(self, exploration_weight=1.0):
        """!
        @brief Select the best child node using UCT (Upper Confidence bounds applied to Trees)
        @param exploration_weight Weight for exploration vs exploitation trade-off
        @return Best child node according to UCT formula
        """
        # Example: UCT formula
        best_score = float('-inf')
        best = None
        for child in self.children:
            if child.visits == 0:
                score = float('inf')
            else:
                exploitation = child.total_reward / child.visits
                exploration = exploration_weight * math.sqrt(math.log(self.visits) / child.visits)
                score = exploitation + exploration
            if score > best_score:
                best_score = score
                best = child
        return best

    def backpropagate(self, reward):
        """!
        @brief Backpropagate the reward up the tree to all ancestors
        @param reward Reward value to propagate
        """
        self.visits += 1
        self.total_reward += reward
        if self.parent:
            self.parent.backpropagate(reward)

File: rl/mcts/mcts_fast/mcts_node.cpp
#include "mcts_node.h"
#include <algorithm>
#include <cmath>
#include <limits>
#include <sstream>

namespace mcts_fast {

MCTSNode::MCTSNode(
    const ProblemState& state,
    std::shared_ptr<MCTSNode> parent,
    const std::optional<ActionTuple>& action,
    int depth
) : state_(state), parent_(parent), action_(action), depth_(depth), 
    visits_(0), total_reward_(0.0), actions_cached_(false) {
}

void MCTSNode::addChild(std::shared_ptr<MCTSNode> child) {
    children_.push_back(child);
}

bool MCTSNode::isRoot() const {
    return parent_ == nullptr;
}

bool MCTSNode::isTerminal() const {
    return state_.isTerminal();
}

bool MCTSNode::isFullyExpanded() const {
    return getUntriedActions().empty();
}

std::vector<ActionTuple> MCTSNode::getUntriedActions() const {
    if (!actions_cached_) {
        cached_possible_actions_ = state_.getPossibleActions();
        actions_cached_ = true;
    }
    
    std::vector<ActionTuple> untried_actions;
    
    // If root node has a specific action without parameters
    if (isRoot() && action_.has_value() && action_->second.empty()) {
        const std::string& action_name = action_->first;
        auto all_params = state_.enumerateValidParams(action_name);
        
        std::vector<ActionParams> tried_params;
        for (const auto& child : children_) {
            if (child->action_.has_value()) {
                tried_params.push_back(child->action_->second);
            }
        }
        
        for (const auto& param : all_params) {
            if (std::find(tried_params.begin(), tried_params.end(), param) == tried_params.end()) {
                untried_actions.emplace_back(action_name, param);
            }
        }
        return untried_actions;
    }
    
    // Normal behavior for other nodes
    std::vector<ActionTuple> tried_actions;
    for (const auto& child : children_) {
        if (child->action_.has_value()) {
            tried_actions.push_back(child->action_.value());
        }
    }
    
    for (const auto& action : cached_possible_actions_) {
        if (std::find(tried_actions.begin(), tried_actions.end(), action) == tried_actions.end()) {
            untried_actions.push_back(action);
        }
    }
    
    return untried_actions;
}

std::shared_ptr<MCTSNode> MCTSNode::expand(const ActionTuple& action) {
    ProblemState new_state = state_.copy();
    new_state.applyAction(action.first, action.second);
    
    auto child_node = std::make_shared<MCTSNode>(
        new_state, shared_from_this(), action, depth_ + 1);
    addChild(child_node);
    
    return child_node;
}

std::shared_ptr<MCTSNode> MCTSNode::bestChild(double exploration_weight) const {
    if (children_.empty()) {
        return nullptr;
    }
    
    double best_score = std::numeric_limits<double>::lowest();
    std::shared_ptr<MCTSNode> best_child = nullptr;
    
    for (const auto& child : children_) {
        double score;
        if (child->visits_ == 0) {
            score = std::numeric_limits<double>::infinity();
        } else {
            double exploitation = child->total_reward_ / child->visits_;
            double exploration = exploration_weight * 
                std::sqrt(std::log(visits_) / child->visits_);
            score = exploitation + exploration;
        }
        
        if (score > best_score) {
            best_score = score;
            best_child = child;
        }
    }
    
    return best_child;
}

void MCTSNode::backpropagate(double reward) {
    visits_++;
    total_reward_ += reward;
    
    if (parent_) {
        parent_->backpropagate(reward);
    }
}

double MCTSNode::getUCTValue(double exploration_weight, int parent_visits) const {
    if (visits_ == 0) {
        return std::numeric_limits<double>::infinity();
    }
    
    double exploitation = total_reward_ / visits_;
    double exploration = exploration_weight * std::sqrt(std::log(parent_visits) / visits_);
    return exploitation + exploration;
}

std::string MCTSNode::toString() const {
    std::ostringstream oss;
    oss << "MCTSNode(depth=" << depth_ << ", visits=" << visits_ 
        << ", total_reward=" << total_reward_;
    
    if (action_.has_value()) {
        oss << ", action=" << action_->first << " [";
        for (size_t i = 0; i < action_->second.size(); ++i) {
            if (i > 0) oss << ", ";
            oss << action_->second[i];
        }
        oss << "]";
    }
    
    oss << ", children=" << children_.size() << ")";
    return oss.str();
}

} // namespace mcts_fast